### PR TITLE
Make RCLASS_EXT(c)->subclasses a doubly linked list

### PR DESCRIPTION
**Pull request by eightbitraptor**
When sweeping a class, we have to remove it from it's superclass's subclass list and remove all it's subclasses. the `rb_classext_t` retains a double pointer to the classes superclass's list of subclasses in order to facilitate this.

This means that when we free a class, we have to modify the superclass to remove/reset that link. This is fine when RVARGC is disabled as the `rb_classext_t` is not stored in the Ruby heap.

with RVARGC enabled the `rb_classext_t` lives in the Ruby heap, which means that when we're sweep and compacting, we need to modify the heap in ways that we aren't expecting.

This PR removes the need to modify the Ruby heap when freeing classes by keeping the entire subclass list in the C heap and making it a doubly linked list with a dedicated list head.

### CODE REVIEW

File: class.c [hunk 1]
```diff
@@ -63,70 +67,96 @@ rb_module_add_to_subclasses_list(VALUE module, VALUE iclass)
...
     /* do not be tempted to simplify this loop into a for loop, the order of
        operations is important here if `f` modifies the linked list */
     while (cur) {
 	VALUE curklass = cur->klass;
-	cur = cur->next;
+        cur = cur->next;
         // do not trigger GC during f, otherwise the cur will become
         // a dangling pointer if the subclass is collected
-	f(curklass, arg);
+        if (curklass) {
```
Comment (peterzhu2118) [inline class.c:157]: We shouldn't need this check as long as we skip the first dummy head (all nodes thereafter should have `curklass` set).

File: class.c [hunk 2]
```diff
@@ -965,17 +995,19 @@ rb_include_module(VALUE klass, VALUE module)
...
+                    check_class = RCLASS_SUPER(check_class);
                 }
-                check_class = RCLASS_SUPER(check_class);
-            }
 
-            if (do_include) {
-                include_modules_at(iclass->klass, RCLASS_ORIGIN(iclass->klass), module, TRUE);
+                if (do_include) {
+                    include_modules_at(iclass->klass, RCLASS_ORIGIN(iclass->klass), module, TRUE);
+                }
```
Comment (peterzhu2118) [inline class.c:1010]: Why is this change necessary?

File: class.c [hunk 3]
```diff
@@ -1206,17 +1238,19 @@ rb_prepend_module(VALUE klass, VALUE module)
...
+                    // backfill an origin iclass to handle refinements and future prepends
+                    rb_id_table_foreach(RCLASS_M_TBL(iclass->klass), clear_module_cache_i, (void *)iclass->klass);
+                    RCLASS_M_TBL(iclass->klass) = klass_m_tbl;
+                    VALUE origin = rb_include_class_new(klass_origin, RCLASS_SUPER(iclass->klass));
+                    RCLASS_SET_SUPER(iclass->klass, origin);
+                    RCLASS_SET_INCLUDER(origin, RCLASS_INCLUDER(iclass->klass));
+                    RCLASS_SET_ORIGIN(iclass->klass, origin);
+                    RICLASS_SET_ORIGIN_SHARED_MTBL(origin);
+                }
+                include_modules_at(iclass->klass, iclass->klass, module, FALSE);
```
Comment (peterzhu2118) [inline class.c:1252]: Why is this change necessary?

File: gc.c [hunk 4]
```diff
@@ -3131,13 +3131,11 @@ obj_free(rb_objspace_t *objspace, VALUE obj)
...
-	    }
-	    else {
-		rb_class_detach_subclasses(obj);
-	    }
-	    RCLASS_SUBCLASSES(obj) = NULL;
+            if (BUILTIN_TYPE(obj) == T_MODULE) {
+                rb_class_detach_module_subclasses(obj);
+            } else {
+                rb_class_remove_subclass_head(obj);
+            }
```
Comment (peterzhu2118) [inline gc.c:3138]: This removes every node of the linked list for the superclass. The superclass should only be remove the dummy head.

File: class.c (hunk 2 above)
Comment (eightbitraptor) [reply class.c:1010]: I added checks for the presence of `rb_subclass_entry->klass` into the places that walk the subclass list manually, as a way of effectively skipping the dummy head. My assumption is that having an `rb_subclass_entry` with no `klass` is not possible except for the dummy head.

File: class.c [hunk 5]
```diff
@@ -63,61 +68,96 @@ rb_module_add_to_subclasses_list(VALUE module, VALUE iclass)
...
+        head = ZALLOC(rb_subclass_entry_t);
+        RCLASS_SUBCLASSES(module) = head;
+    }
+    entry->next = head->next;
+    entry->prev = head;
+
+    if (head->next) {
+        head->next->prev = entry;
     }
+    head->next = entry;
```
Comment (peterzhu2118) [inline class.c:85]: I think this code can be extracted into a function since it's the same as the one in `rb_class_subclass_add`.

File: class.c [hunk 6]
```diff
@@ -63,61 +68,96 @@ rb_module_add_to_subclasses_list(VALUE module, VALUE iclass)
...
-	*prev = next;
+        if (prev) {
+            prev->next = next;
+        }
 	if (next) {
-	    RCLASS_MODULE_SUBCLASSES(next->klass) = prev;
+            next->prev = prev;
 	}
 
 	xfree(entry);
```
Comment (peterzhu2118) [inline class.c:140]: I think this can also be extracted to a function.

File: class.c [hunk 7]
```diff
@@ -63,61 +68,96 @@ rb_module_add_to_subclasses_list(VALUE module, VALUE iclass)
...
+    // if we have a subclasses list, then the head is a placeholder with no valid
+    // class. So ignore it and use the next element in the list (if one exists)
+    if (cur && !cur->klass) {
+        if (cur->next) {
+            cur = cur->next;
+        }
+        else {
+            cur = NULL;
+        }
+    }
```
Comment (peterzhu2118) [inline class.c:160]: I think this can just be:

```suggestion
    if (cur) {
        RUBY_ASSERT(!cur->klass);
        cur = cur->next;
    }
```

File: gc.c [hunk 8]
```diff
@@ -3131,13 +3131,7 @@ obj_free(rb_objspace_t *objspace, VALUE obj)
             rb_id_table_free(RCLASS_CVC_TBL(obj));
 	}
 	if (RCLASS_SUBCLASSES(obj)) {
```
Comment (peterzhu2118) [inline gc.c:3133]: `rb_class_remove_subclass_head` will already check that `RCLASS_SUBCLASSES` is present. We don't need to check it here.

File: gc.c [hunk 9]
```diff
@@ -3304,8 +3298,7 @@ obj_free(rb_objspace_t *objspace, VALUE obj)
 	    rb_id_table_free(RCLASS_CALLABLE_M_TBL(obj));
 	}
 	if (RCLASS_SUBCLASSES(obj)) {
```
Comment (peterzhu2118) [inline gc.c:3300]: Ditto.

File: class.c [hunk 10]
```diff
@@ -37,87 +37,118 @@
...
+        head->next = entry;
+
+        if (BUILTIN_TYPE(super) == T_MODULE) {
+            RUBY_ASSERT(BUILTIN_TYPE(klass) == T_ICLASS);
+            RCLASS_MODULE_SUBCLASS_ENTRY(klass) = entry;
+        }
+        else {
+            RUBY_ASSERT(BUILTIN_TYPE(super) == T_CLASS);
+            RCLASS_SUBCLASS_ENTRY(klass) = entry;
+        }
```
Comment (peterzhu2118) [inline class.c:69]: Why not just keep these in their respective functions (`rb_class_subclass_add` and `rb_module_add_to_subclasses_list`)?

Comment (eightbitraptor) [reply class.c]: I had thought that adding the entry to the front of the list and updating the list pointer were closely related enough that they should be done in the same place, but on reflection I think it makes the code neater to update the list separately and then update the pointer in the original functions. I'll push that up as a change.
